sample_grabber.c

```diff
@@ -299,14 +300,14 @@ static void* file_writer(void* pc_ptr){
     filename_ext = strrchr(output_filename, '.');
     if (filename_ext) {
       basename_len = filename_ext - output_filename;
-      if (sizeof(filename) - basename_len < 22)
-        basename_len = sizeof(filename) - 22; /* leave room for the date */
-      strncpy(filename, output_filename, basename_len);
-      filename[basename_len] = 0;
     } else {
-      strncpy(filename, output_filename, sizeof(filename));
+      basename_len = strlen(output_filename);
       filename_ext = "";
     }
+    if (sizeof(filename) - basename_len < 22)
+        basename_len = sizeof(filename) - 22; /* leave room for the date */
+    strncpy(filename, output_filename, basename_len);
+    filename[basename_len] = 0;
```

Comment:
I don't understand.  I get the length of the string depending on whether it has an extension of not. In either case, I make sure this length + the time string fit in filename.  If its too long, I shorten its length to fit the time and truncate part of the filename.  Then I copy.